src/main/java/org/elasticsearch/script/expression/FieldDataValueSource.java

```diff
@@ -49,13 +48,24 @@ public FunctionValues getValues(Map context, LeafReaderContext leaf) throws IOEx
     }
 
     @Override
-    public boolean equals(Object other) {
-        return fieldData.equals(other);
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        FieldDataValueSource that = (FieldDataValueSource) o;
+
+        return !(fieldData != null ? !fieldData.equals(that.fieldData) : that.fieldData != null);
+
     }
 
     @Override
     public int hashCode() {
-        return fieldData.hashCode();
+        return fieldData != null ? fieldData.hashCode() : 0;
```

Comment:
We don't need to check for null here? fieldData cannot be null (add an assertion in the ctor?)